### PR TITLE
Update Pre-Commit hooks

### PR DESCRIPTION
**Pull request by SaschaCowley**
### Link to issue number:

None

### Summary of the issue:

Continuing work on our pre-commit hooks, there are still several we don't use but could.

### Description of user facing changes

None.

### Description of development approach

Tested pre-commit hooks available at <https://github.com/pre-commit/pre-commit-hooks/>. Added those that work as expected and do not require major changes to our code base. Added:

- Automatic checking of yaml, toml and XML syntax
- Checks that no Python debug statements are included in commits
- Remove BOM on Python, C/C++ and Batch files, if present
- Ensure that unit test files are named according to unittest rules.
- Ensure that any include VCS links are permalinks.

### Testing strategy:

Ran the tests individually with `pre-commit run <hook-id> -a` and all together with `pre-commit run -a`.

### Known issues with pull request:

The following are good candidates for inclusion:

- [check-builtin-literals](https://github.com/pre-commit/pre-commit-hooks/?tab=readme-ov-file#check-builtin-literals): We have a lot of violations of this one
- [check-docstring-first](https://github.com/pre-commit/pre-commit-hooks/?tab=readme-ov-file#check-docstring-first): We have a lot of violations of this one
- [check-illegal-windows-names](https://github.com/pre-commit/pre-commit-hooks/?tab=readme-ov-file#check-illegal-windows-names): Not yet in a tagged version.
- [check-added-large-files](https://github.com/pre-commit/pre-commit-hooks/?tab=readme-ov-file#check-added-large-files): We have several files that are larger than the default limit of 500kb. It is not clear what our cut-off should be, and it doesn't seem as though we can have different sizes for different file types.
  - The biggest culprits seem to be translations, with the largest (Simplified Chinese (PRC)) character descriptions being 3.21MiB.
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Enhanced pre-commit hooks for improved code quality and style enforcement.
  - Added checks for various file types including TOML, YAML, XML, and VCS permalinks.

- **Improvements**
  - Modified existing test hook to enforce unittest standards, refining the validation process for Python files.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -23,10 +23,34 @@ repos:
...
     - id: check-case-conflict
     - id: check-merge-conflict
+    - id: debug-statements
     - id: trailing-whitespace
       types_or: [python, c, c++, batch, markdown]
     - id: end-of-file-fixer
       types_or: [python, c, c++, batch, markdown]
+    - id: fix-byte-order-marker
+      types_or: [python, c, c++, batch, markdown]
+    - id: check-toml
```
Comment (seanbudd) [inline .pre-commit-config.yaml:33]: Do we have any TOML?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -23,10 +23,34 @@ repos:
...
+    - id: check-yaml
+    - id: check-xml
+    - id: check-vcs-permalinks
```
Comment (seanbudd) [inline .pre-commit-config.yaml:36]: Could you add a comment explaining what this does, it's not immediately clear

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -23,10 +23,34 @@ repos:
       exclude: source/comInterfaces/_944DE083_8FB8_45CF_BCB7_C477ACB2F897_0_1_0.py
     - id: check-case-conflict
     - id: check-merge-conflict
+    - id: debug-statements
     - id: trailing-whitespace
       types_or: [python, c, c++, batch, markdown]
     - id: end-of-file-fixer
       types_or: [python, c, c++, batch, markdown]
+    - id: fix-byte-order-marker
```
Comment (seanbudd) [inline .pre-commit-config.yaml:31]: Could you add a comment explaining this one

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -23,10 +23,34 @@ repos:
       exclude: source/comInterfaces/_944DE083_8FB8_45CF_BCB7_C477ACB2F897_0_1_0.py
     - id: check-case-conflict
     - id: check-merge-conflict
+    - id: debug-statements
```
Comment (seanbudd) [inline .pre-commit-config.yaml:26]: Same here re comment

File: .pre-commit-config.yaml (hunk 1 above)
Comment (seanbudd) [reply .pre-commit-config.yaml:33]: Oh right, the new pyproject.toml

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -21,12 +21,48 @@ repos:
     - id: check-ast
       # File uses encoding not supported in Linux
       exclude: source/comInterfaces/_944DE083_8FB8_45CF_BCB7_C477ACB2F897_0_1_0.py
+    # Checks for filenames that will conflict on case insensative filesystems (most windows filesystems, most of the time)
```
[the remark this answers is not in the document]
Comment (SaschaCowley) [reply .pre-commit-config.yaml:30]: Windows isn't a file system. And NTFS supports case sensitivity but it's off by default.

[the remark this answers is not in the document]
Comment (LeonarddeR) [reply .pre-commit-config.yaml:30]: That makes sense. I would probably avoid using the word "most" twice.

[the remark this answers is not in the document]
Comment (XLTechie) [reply .pre-commit-config.yaml:30]: I can't let a conversation this hilariously pedantic go past without sticking my two bits in.
Note: changed spelling of "insensitive", and capitalized "Windows".

```suggestion
    # Checks for filenames that will conflict on case insensitive filesystems (the majority of Windows filesystems, most of the time)
```

[the remark this answers is not in the document]
Comment (SaschaCowley) [reply .pre-commit-config.yaml:30]: Hmm yes I apparently was in a very pedantic mood on Friday haha :)

File: test.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+Test link: https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_vcs_permalinks.py
```
Comment (seanbudd) [inline test.txt:1]: please remove this file

**Review by seanbudd — APPROVED** [review]

**Pull request merged**